qadence_embeddings/embedding.py

```
@@ -0,0 +1,116 @@
+        self._dtype: DTypeLike = None
+        self.time_dependent_vars: list[str] = []
+        self._device = device
+        self._time_vars_identified = False
+
+    @property
+    def root_param_names(self) -> list[str]:
+        return list(self.vparams.keys()) + self.fparam_names
+
+    def embed_all(
+        self,
+        inputs: dict[str, ArrayLike],
+    ) -> dict[str, ArrayLike]:
+        """The standard embedding of all intermediate and leaf parameters.
+        Include the root_params, i.e., the vparams and fparams original values
+        to be reused in computations.
+        """
+        for intermediate_or_leaf_var, engine_callable in self.var_to_call.items():
+            # We mutate the original inputs dict and include intermediates and leaves.
+            if not self._time_vars_identified:
+                # we do this only on the first embedding call
+                if self.tparam_name and any(
+                    [
+                        p in [self.tparam_name] + self.time_dependent_vars
+                        for p in engine_callable.abstract_args
+                    ]  # we check if any parameter in the callables args is time
+                    # or depends on an intermediate variable which itself depends on time
+                ):
+                    self.time_dependent_vars.append(intermediate_or_leaf_var)
+                    # we remember which parameters depend on time
+            inputs[intermediate_or_leaf_var] = engine_callable(inputs)
+        self._time_vars_identified = True
+        return inputs
+
+    def reembed_time(
+        self,
+        embedded_params: dict[str, ArrayLike],
+        tparam_value: ArrayLike,
+    ) -> dict[str, ArrayLike]:
+        """Receive already embedded params containing intermediate and leaf parameters
+        and recalculate the those which are dependent on the time parameter using the new value
+        `tparam_value`.
+        """
+        assert self.tparam_name is not None
```

The semantics here are that `reembed_time` is always called after the initial embedding and therefore `self.tparam_name` is set. I would raise an exception here with a proper message.